spark/src/main/java/org/apache/iceberg/spark/data/SparkValueReaders.java

```diff
@@ -285,4 +291,36 @@ protected void set(InternalRow struct, int pos, Object value) {
       }
     }
   }
+
+  static class UnionReader implements ValueReader<InternalRow> {
+    private final ValueReader[] readers;
+
+    // TODO: need to make use of structSchema
+    private final Types.StructType structType;
+    private final int numFields;
+
+    private UnionReader(List<ValueReader<?>> readers, Types.StructType structType, Map<Integer, ?> idToConstant) {
+      this.readers = new ValueReader[readers.size()];
+      for (int i = 0; i < this.readers.length; i += 1) {
+        this.readers[i] = readers.get(i);
+      }
+
+      this.structType = structType;
+      this.numFields = idToConstant.size();
+    }
+
+    @Override
+    public InternalRow read(Decoder decoder, Object reuse) throws IOException {
+      InternalRow struct = new GenericInternalRow(readers.length);
+      int index = decoder.readIndex();
+      Object value = this.readers[index].read(decoder, reuse);
+
+      struct.update(index, value);
+      for (int i = 0; i < readers.length && i != index; i += 1) {
```

Comment:
should we do this for all i before we set the value at index? This loop is setting null all fields before the value, but not after...

Reply:
Thanks for catching this. Change to explicitly set all fields to null before setting value at index.